### ISSUE TITLE
compiler: hyper-connection mixing op (mHC residual streams)

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#99]
## Context
Manifold-constrained hyper-connections: the residual is a stack of `hc_mult` parallel streams mixed per layer by `pre`/`post`/`comb` weights, with `comb` projected onto the doubly-stochastic manifold by Sinkhorn-Knopp (floe `deepseek_v4/forward.py::DeepseekV4HyperConnection`; glm53flash `Glm53HyperConnection`).

## Semantics (decode step)
- per layer: `pre`/`post`/`comb` are **data-dependent** — computed from the stream contents by one small linear each (`F.linear(input_norm(flatten(streams)), fn)`), so the Sinkhorn-Knopf projection of `comb` runs **per token**, not at load time (correction of the original assumption)
- `h_in = Σ_h pre_h · stream_h`; block body; `streams’ = comb · streams + post ⊙ body_out` — small per-row chains over `[hc_mult, C]`

## Work
- `hyper_mix` elementwise-ish op family (task per row); or decompose into existing `linear`/elementwise where shapes allow — the einsum chains are rank-2, likely expressible as several small GEMVs with add; decide during lowering
- stream state is just intermediate buffers (workspace), not persistent

## Validation
tiny-config eager block parity; stream contents checked at layer boundaries.

Labels: arch:ops, model:deepseek-v4, model:glm-5.3

## Implementation sketch

**Backend**: Triton, one task per row per layer boundary (elementwise + tiny GEMV chains — no `tl.dot`; the `fn` projection is one `[hc·D] → [hc + hc + hc²]` GEMV, fold into the task or reuse `gemv`).

**Arithmetic** (fp32 throughout, per floe `DeepseekV4HyperConnection.forward`; `hc = hc_mult` is small):
```
flat   = unweighted_rms_norm(flatten(streams))            # [hc·D]
pre_w, post_w, comb_w = linear(flat, fn)                  # data-dependent!
pre    = sigmoid(pre_w·pre_s + pre_b) + eps                # [hc]
post   = 2·sigmoid(post_w·post_s + post_b)                 # [hc], range (0,2)
comb   = softmax((comb_w·comb_s + comb_b).view(hc, hc)) + eps
comb  /= colsum(comb)                                      # Sinkhorn-Knopp:
for _ in iters−1: comb /= rowsum; comb /= colsum           # alternate to doubly-stochastic
h_in   = Σ_h pre_h · streams[:, h]                         # stream collapse
```
**Dtypes**: streams bf16, all mixing math fp32 (Sinkhorn numerics demand it), h_in bf16 out. GLM's `Glm53HyperConnection` is the same op family — one implementation, family attributes.

**Comment by xzyaoi** [comment]
## Implementation: mHC hyper-connection mixing op (`mhc_pre`/`mhc_post`) — branch `issue-99-hyper-connection-mixing`, PR to follow

**Verified, not attested.** The floe-parity gates were run to green in a torch-bearing environment (serving-stack floe venv, torch 2.13 CPU), not merely skipped-and-attested.

### What shipped (6 commits, rebased on current main incl. #114 gdn_conv / #88 rope / #94 slot-table / #106 dsa)
1. `3311b1d` — op kinds `mhc_pre`/`mhc_post` + `RecordingBackend.mhc_pre/mhc_post` capture recorders (family attributes `hc`, `iters`, `eps`, `rms_eps`, `layer`; hazard edges on the stream stack and the data-dependent weight workspace).
2. `bc83279` — task lowering (fn projection FOLDED into the pre task per the `mhc_pre_gemm_sqrsum` fusion precedent; no separate linear op) + fp64 reference-exec bodies.
3. `54f5312` — `_t_mhc_pre`/`_t_mhc_post` Triton device templates.
4. `fbb04d2`+`06fc981`+`7d1b6ae` — 730-line oracle suite + fixes.

### Numerical contract (per decode token, streams [B, hc, C], mix = (2+hc)·hc)
- `flat = streams[b].flatten()`; unweighted RMSNorm `flat / sqrt(mean(flat²) + rms_eps)`
- `logits = fn @ flat` — **F.linear carries NO projection bias**; `base` enters only inside the gates
- split `pre_w | post_w | comb_w` = [`hc` | `hc` | `hc²`]
- gates: `pre = sigmoid(pre_w·pre_s + pre_b)`; `post = 2·sigmoid(post_w·post_s + post_b)`; `comb = softmax(hc²·Sinkhorn_iters(comb_w·comb_s + comb_b) + eps)`
- collapse: `h = Σ_j pre_j · streams_j` (body input, per hc row)
- compose: `streams' = combᵀ @ streams + post ⊙ body_out` (matmul over hc, elementwise post)

### The bug the floe gate caught (why the bare-env suite alone would have shipped it)
My executor body, numpy mirror, and Triton template all added `base` to the raw projection logits **and** inside the gates — double-counting base. The bare-env suite couldn't catch it because the executor and the mirror shared the same misreading; the floe-parity test diverged from logits row 0. Fixed in all three implementations (`06fc981`).

### Test results
- Bare env (numpy-only): **358 passed**, 1 failed — `test_discovery::test_kernel_names_and_order`, **pre-existing on main** (verified by stash).
- floe venv (torch 2.13 CPU): **489 passed**, 3 failed — all three pre-existing on main (2× discovery, 1× vllm, verified by stash).
- mHC module alone, torch env: **23 passed, 1 skipped** (Triton device test — GPU-only).
- 2 floe-parity gates now run to green in the torch env: `test_floe_deepseek_hc_parity`, `test_floe_glm_hc_parity` (DeepseekV4HyperConnection + Glm53HyperConnection vs the executor, all intermediates: post, comb, collapsed h, streams').
- Also fixed en route: `tests/python/conftest.py` pins the repo src tree over pip-installed vkernels (the floe venv installs it editable from the main checkout, which split module identities — two `CaptureError` classes — and hid branch-only subpackages).

**Comment by xzyaoi** [comment]
Superseded by #129 (Stage 2 megakernel, merged): the mHC hyper-connection ops (mhc_pre / mhc_post) landed with the DeepSeek-V4 decode step. OP_MHC_PRE / OP_MHC_POST present in operator_ir.py on main. Closing per the supersession note on PR #124.
